File: answers/leetcode/Min Stack/Min Stack.cpp
//@result 17 / 17 test cases passed. Status: Accepted Runtime: 28 ms Submitted: 0 minutes ago You are here! Your runtime beats 42.34% of cpp submissions.

class MinStack {
public:
	void push(int x) {
		my_stack.push(x);
		if (min_stack.empty() || x <= min_stack.top()) {
			min_stack.push(x);
		}
	}

	void pop() {
		if (! my_stack.empty()) {
			int val = my_stack.top();
			my_stack.pop();
			if (! min_stack.empty() && val == min_stack.top()) {
				min_stack.pop();
			}
		}
	}

	int top() {
		if (! my_stack.empty()) {
			return my_stack.top();
		}
		else {
			return INT_MAX;
		}
	}

	int getMin() {
		if (! min_stack.empty()) {
			return min_stack.top();
		}
		else {
			return INT_MAX;
		}
	}
	std::stack<int> my_stack, min_stack;
};

File: answers/leetcode/Min Stack/Min Stack.java
class MinStack {

    private long min;
    private Stack<Long> stack;

    /** initialize your data structure here. */
    public MinStack() {
        this.min = 0;
        this.stack = new Stack<Long>();
    }

    public void push(int x) {
        long longX = (long) x;
        long newValue = 0;
        if (this.stack.isEmpty()) {
            newValue = 0;
            this.min = longX;
        } else if (this.min > longX) {
            newValue = longX - this.min;
            this.min = longX;
        } else {
            newValue = longX - this.min;
        }
        this.stack.push(newValue);
    }

    public void pop() {
        if (!this.stack.isEmpty()) {
            long top = this.stack.pop();
            if (top < 0) {
                this.min -= top;
            }
        }
    }

    public int top() {
        long top = this.stack.peek();
        if (top >= 0) {
            return (int) (top + this.min);
        } else {
            return (int) this.min;
        }
    }

    public int getMin() {
        return (int) this.min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */